File: src/StaticRing.zig
const std = @import("std");
const expect = std.testing.expect;
const expectEqual = std.testing.expectEqual;
const expectError = std.testing.expectError;

pub fn StaticRing(comptime T: type, capacity: usize) type {
    const BitSet = std.StaticBitSet(capacity);

    return struct {
        const Self = @This();

        const RingNode = struct {
            previous: ?*RingNode = null,
            next: ?*RingNode = null,
            value: T,
        };

        values: [capacity]RingNode = undefined,
        capacity: usize = capacity,
        first: ?*RingNode = null,
        last: ?*RingNode = null,
        free: BitSet = BitSet.initFull(),
        size: usize = 0,

        /// Appends an element to the end of the list.
        pub fn append(ring: *Self, value: T) !*RingNode {
            if (ring.size == capacity) return error.OutOfMemory;
            const index = ring.free.findFirstSet().?;
            ring.free.unset(index);

            const new_node = &ring.values[index];
            new_node.value = value;

            if (ring.size == 0) {
                new_node.previous = new_node;
                new_node.next = new_node;
                ring.first = new_node;
                ring.last = new_node;
            } else {
                new_node.previous = ring.last;
                new_node.next = ring.first;
                ring.last.?.next = new_node;
                ring.first.?.previous = new_node;
                ring.last = new_node;
            }

            ring.size += 1;
            return new_node;
        }

        /// Prepends an element to the start of the list.
        pub fn prepend(ring: *Self, value: T) !*RingNode {
            if (ring.size == capacity) return error.OutOfMemory;
            const index = ring.free.findFirstSet().?;
            ring.free.unset(index);

            const new_node = &ring.values[index];
            new_node.value = value;

            if (ring.size == 0) {
                new_node.previous = new_node;
                new_node.next = new_node;
                ring.first = new_node;
                ring.last = new_node;
            } else {
                new_node.next = ring.first;
                new_node.previous = ring.last;

                ring.first.?.previous = new_node;
                ring.last.?.next = new_node;
                ring.first = new_node;
            }
            ring.size += 1;
            return new_node;
        }

        /// Inserts an element after a given node.
        pub fn insertAfter(ring: *Self, node: *RingNode, value: T) !*RingNode {
            if (ring.size == capacity) return error.OutOfMemory;
            const index = ring.free.findFirstSet().?;
            ring.free.unset(index);

            const new_node = &ring.values[index];
            new_node.value = value;

            new_node.previous = node;
            new_node.next = node.next;

            if (node.next) |next_node| {
                next_node.previous = new_node;
            }
            node.next = new_node;

            if (node == ring.last.?) {
                ring.last = new_node;
                new_node.next = ring.first;
                ring.first.?.previous = new_node;
            }
            ring.size += 1;
            return new_node;
        }

        /// Removes a node from the list.
        pub fn remove(ring: *Self, node: *RingNode) void {
            if (ring.size == 0) return;
            if (ring.size == 1) {
                ring.first = null;
                ring.last = null;
            } else {
                if (node.previous) |prev_node| {
                    prev_node.next = node.next;
                }
                if (node.next) |next_node| {
                    next_node.previous = node.previous;
                }
                if (ring.first == node) {
                    ring.first = node.next;
                }
                if (ring.last == node) {
                    ring.last = node.previous;
                }
            }

            const index =
                (@intFromPtr(node) - @intFromPtr(&ring.values[0])) / @sizeOf(RingNode);
            ring.free.set(index);
            ring.size -= 1;
            ring.values[index].next = null;
            ring.values[index].previous = null;
        }

        /// Pops the last element from the list and returns its value.
        pub fn pop(ring: *Self) ?T {
            if (ring.size == 0) return null;
            const last_node = ring.last.?;
            const value = last_node.value;
            ring.remove(last_node);
            return value;
        }

        /// Pops the first element from the list and returns its value.
        pub fn popFirst(ring: *Self) ?T {
            if (ring.size == 0) return null;
            const first_node = ring.first.?;
            const value = first_node.value;
            ring.remove(first_node);
            return value;
        }

        /// Returns the number of elements in the list.
        pub fn length(ring: *Self) usize {
            return ring.size;
        }

        pub fn reset(ring: *Self) void {
            ring.free.setRangeValue(.{ .start = 0, .end = capacity }, true);
            ring.first = null;
            ring.last = null;
        }
    };
}

test "RingDoublyLinkedList" {
    const List = StaticRing(u32, 4);
    var list = List{};

    const n1 = try list.append(10);
    try expectEqual(n1.value, 10);

    try expectEqual(list.first, n1);
    try expectEqual(list.last, n1);

    const n2 = try list.append(20);
    const n3 = try list.prepend(5);
    const n4 = try list.append(30);

    try expectEqual(list.size, 4);
    try expectError(error.OutOfMemory, list.append(40));

    try expectEqual(list.first.?.previous, list.last);
    try expectEqual(list.last.?.next, list.first);

    try expectEqual(list.first, n3);
    try expectEqual(n3.next, n1);
    try expectEqual(n1.next, n2);
    try expectEqual(n2.next, n4);
    try expectEqual(n4.next, list.first);

    list.remove(n3);
    try expectEqual(list.size, 3);
    try expectEqual(list.first.?.value, 10);
    try expectEqual(list.first.?.previous, list.last);
    try expectEqual(list.last.?.next, list.first);
}

File: ts/src/staticRing.ts
import { nullthrows as nt } from "./nullthrows";

export class StaticRing {
  private readonly values: Int32Array;
  private readonly next: Int32Array;
  private readonly prev: Int32Array;
  private readonly inUse: Uint8Array;
  private readonly freeStack: Int32Array;
  private freeTop: number;

  private firstIndex = -1;
  private lastIndex = -1;
  size = 0;

  constructor(public readonly capacity: number) {
    if (!Number.isInteger(capacity) || capacity <= 0) {
      throw new RangeError("StaticRing: capacity must be positive integer");
    }
    this.values = new Int32Array(capacity);
    this.next = new Int32Array(capacity).fill(-1);
    this.prev = new Int32Array(capacity).fill(-1);
    this.inUse = new Uint8Array(capacity);
    this.freeStack = new Int32Array(capacity);
    this.freeTop = capacity;
    for (let i = 0; i < capacity; i += 1) {
      this.freeStack[i] = capacity - 1 - i;
    }
  }

  private allocNode(): number {
    if (this.freeTop === 0) {
      throw new RangeError("StaticRing overflow");
    }
    const index = nt(this.freeStack[--this.freeTop]);
    this.inUse[index] = 1;
    this.next[index] = -1;
    this.prev[index] = -1;
    return index;
  }

  private freeNode(index: number): void {
    this.inUse[index] = 0;
    this.next[index] = -1;
    this.prev[index] = -1;
    this.freeStack[this.freeTop++] = index;
  }

  get first(): number {
    return this.firstIndex;
  }

  get last(): number {
    return this.lastIndex;
  }

  append(value: number): number {
    const index = this.allocNode();
    this.values[index] = value;

    if (this.size === 0) {
      this.next[index] = index;
      this.prev[index] = index;
      this.firstIndex = index;
      this.lastIndex = index;
    } else {
      const last = this.lastIndex;
      const first = this.firstIndex;
      this.next[last] = index;
      this.prev[index] = last;
      this.next[index] = first;
      this.prev[first] = index;
      this.lastIndex = index;
    }

    this.size += 1;
    return index;
  }

  prepend(value: number): number {
    const index = this.allocNode();
    this.values[index] = value;

    if (this.size === 0) {
      this.next[index] = index;
      this.prev[index] = index;
      this.firstIndex = index;
      this.lastIndex = index;
    } else {
      const first = this.firstIndex;
      const last = this.lastIndex;
      this.prev[first] = index;
      this.next[index] = first;
      this.prev[index] = last;
      this.next[last] = index;
      this.firstIndex = index;
    }

    this.size += 1;
    return index;
  }

  insertAfter(nodeIndex: number, value: number): number {
    const index = this.allocNode();
    this.values[index] = value;

    const nextIndex = nt(this.next[nodeIndex]);
    this.next[nodeIndex] = index;
    this.prev[index] = nodeIndex;
    this.next[index] = nextIndex;
    this.prev[nextIndex] = index;

    if (nodeIndex === this.lastIndex) {
      this.lastIndex = index;
    }

    this.size += 1;
    return index;
  }

  remove(nodeIndex: number): void {
    if (this.size === 1) {
      this.firstIndex = -1;
      this.lastIndex = -1;
      this.freeNode(nodeIndex);
      this.size = 0;
      return;
    }

    const prevIndex = nt(this.prev[nodeIndex]);
    const nextIndex = nt(this.next[nodeIndex]);

    this.next[prevIndex] = nextIndex;
    this.prev[nextIndex] = prevIndex;

    if (nodeIndex === this.firstIndex) {
      this.firstIndex = nextIndex;
    }
    if (nodeIndex === this.lastIndex) {
      this.lastIndex = prevIndex;
    }

    this.freeNode(nodeIndex);
    this.size -= 1;
  }

  pop(): number | null {
    if (this.size === 0) return null;
    const value = this.values[this.lastIndex];
    this.remove(this.lastIndex);
    return value ?? null;
  }

  popFirst(): number | null {
    if (this.size === 0) return null;
    const value = this.values[this.firstIndex];
    this.remove(this.firstIndex);
    return value ?? null;
  }

  length(): number {
    return this.size;
  }

  valueOf(index: number): number {
    return this.values[index]!;
  }

  nextOf(index: number): number {
    return this.next[index]!;
  }

  prevOf(index: number): number {
    return this.prev[index]!;
  }

  reset(): void {
    this.size = 0;
    this.firstIndex = -1;
    this.lastIndex = -1;
    for (let i = 0; i < this.values.length; i += 1) {
      if (this.inUse[i]) {
        this.inUse[i] = 0;
        this.next[i] = -1;
        this.prev[i] = -1;
      }
      this.freeStack[i] = this.values.length - 1 - i;
    }
    this.freeTop = this.values.length;
  }
}
